text/1011-query-rewrite.rst

```diff
@@ -0,0 +1,322 @@
+statement and can refer to the set as ``__result__``, for example::
+
+    POLICY ON UPDATE DO AFTER (
+        FOR obj IN __result__ UNION (
+        INSERT AuditLog {
+            text := 'updated ' ++ obj.name
+        })
+    )
+
+    # or
+
+    POLICY ON INSERT DO AFTER (
+        INSERT EventLog {
+            text := 'inserted ' ++ count(__result__) ++ ' things'
+        }
+    )
+
+The ``PERMIT`` and ``RESTRICT TO`` actions are *filtering* actions and define
+which elements of the object set are visible in a query.  If at least one
+filtering policy is defined for a type, then all elements become invisible by
+default and must be made visible by at least one ``PERMIT`` policy.
+
+The filtering actions are defined as follows:
+
+- ``PERMIT`` -- add all objects, for which the expression evaluates to a
+  set containing at least one ``true`` value, to the set of visible objects.
+  ``PERMIT`` policies are combined using the ``OR`` operator.
+
+- ``RESTRICT TO`` -- restrict the set of visible objects to only objects for
+  which the expression evaluates to a set containing at least one ``true``
+  value.  ``RESTRICT`` policies are combined using the ``AND`` operator and
+  are applied *after* all ``PERMIT`` policies.
+
+The ``CHECK`` policy action is a *validation* action and is applied to
+``INSERT`` and ``UPDATE`` statements.  An object modification will be
+allowed only if the ``<element-expr>`` returns a set containing at least one
+``true`` value, otherwise an error will be raised and the statement execution
+aborted.  Note that in ``UPDATE`` the expression is applied to the object will
+all updates applied, not the original object.
+
+The ``DO AFTER`` and ``DO ON COMMIT`` are *trigger* actions and are applied
+to the result of a statement *as a whole*.  The result set is available in
+the ``<set-expr>`` expression as ``__result__``.  ``DO AFTER`` action
+is applied directly onto the result of an affected statement, whereas
+``DO ON COMMIT`` is applied shortly after the transaction containing the
+affected statement has been successfully committed.  Thus, ``DO AFTER`` is
+suitable for actions that are themselves transactional in nature and can be
+rolled back together with the relevant statement, such as writing into an
+audit log set.  ``DO ON COMMIT``, on the other hand, is designed for actions
+that generate non-transactional external side effects, such as emitting an
+event on an external message bus.  Unlike other actions, errors raised in
+``DO ON COMMIT`` action do not affect the original statement and will only
+be logged.
+
+Allowed statement/action combination are as follows:
+
++--------------+--------+--------+--------+--------+
+|              | SELECT | INSERT | UPDATE | DELETE |
++--------------+--------+--------+--------+--------+
+| PERMIT       |  yes   |   n/a  |  yes   |  yes   |
++--------------+--------+--------+--------+--------+
+| RESTRICT     |  yes   |   n/a  |  yes   |  yes   |
```

Comment:
Hm.  Are you suggesting we should convert `PERMIT/RESTRICT TO` into a `CHECK` action automatically?  It's possible, but might also be confusing.  Because these are permissions and not constraints, it's plausible to have no read access, but have write access, so if you automatically derive write permissions from read permissions, you would have to make an exception, like:

```
# restrict reads to owner
policy on select permit (.owner = global user)

# but allow writes from others
policy on insert check (true)
```

Though I now think there should also be a `permit/restrict` dualism for `check` actions also, so need to think about this more.

Reply:
> Are you suggesting we should convert PERMIT/RESTRICT TO into a CHECK action automatically? It's possible, but might also be confusing. Because these are permissions and not constraints, it's plausible to have no read access, but have write access, so if you automatically derive write permissions from read permissions, you would have to make an exception

Sort of. Although, what I propose is:
```
policy on select, insert permit (.owner = global user)
```
So there is no conflict. As you can remove `insert` in `policy` statement and set another policy for inserts. I think  this is still more straightforward that duplicating the same for `check`.

Reply:
The problem is `update`, because there you need to both filter existing objects _and_ check the updated data, and those conditions might not necessarily be the same.